cfour/default.nix

```diff
@@ -0,0 +1,66 @@
+{ stdenv, lib, makeWrapper, gfortran, requireFile, writeTextFile, python3, blas-i8, lapack,
+  mpi ? null
+}:
+# Make sure to have a correct version of BLAS.
+assert
+  lib.asserts.assertMsg
+  (blas-i8.isILP64)
+  "A 64 bit integer implementation of BLAS is required.";
+
+# Only works with OpenMPI
+assert
+  lib.asserts.assertMsg
+  (mpi == null || mpi.pname == "openmpi")
+  "If MPI is to be used for CFour, use OpenMPI!";
+let
+  useMPI = mpi != null;
+  useMKL64 = blas-i8.passthru.implementation == "mkl";
```

Comment:
I wonder if that should be done via `useMKL64`.  The blas/lapack switching mechanism in nixpkgs should make it easier by using overrides like this (https://nixos.org/manual/nixpkgs/stable/#sec-overlays-alternatives-blas-lapack):
```
callPackage ./cfour {
    blas = super.blas.override { blasProvider = super.mkl; };
    lapack = super.lapack.override { lapackProvider = super.mkl; };
};
```
That would save you a lot of code in the derivation where you make the distinctions between MKL/non-MKL.

Reply:
Yes, this actually a very convenient thing to have and I would like to use it more often. I enabled it only in the places, where this mechanism unfortunately fails. It seems common, that QC codes try to detect if they link against MKL and use some special flags then. The NixPkgs mechanism does not seem to be completely transparent and e.g. CFour figures out, that it is not a generic `-lblas` that it is seeing but actually MKL under the hood. It tries to use MKL flags then and fails. Therefore, I usually unwrap MKL in those cases. They are actually (and unfortunately) by intention.

Reply:
> It seems common, that QC codes try to detect if they link against MKL and use some special flags then. 

Unfortunately Molcas does the same thing.

> The NixPkgs mechanism does not seem to be completely transparent and e.g. CFour figures out, that it is not a generic -lblas that it is seeing but actually MKL under the hood.

You can always look for MKL specific symbols/functions in the library to figure out with implementation it is. The `nixpkgs` switching mechanism does not try to hide it but rather provides the linker flags in a discrete manner.